ledger/src/blockstore.rs

```diff
@@ -3281,21 +3262,46 @@ impl Blockstore {
                 })
                 .collect();
         let data_shreds = data_shreds?;
-        let last_shred = data_shreds.last().unwrap();
-        assert!(last_shred.data_complete() || last_shred.last_in_slot());
 
-        let deshred_payload = Shredder::deshred(&data_shreds).map_err(|e| {
-            BlockstoreError::InvalidShredData(Box::new(bincode::ErrorKind::Custom(format!(
-                "Could not reconstruct data block from constituent shreds, error: {e:?}"
-            ))))
-        })?;
+        let mut entries = vec![];
+        for (start_index, end_index) in completed_ranges.iter() {
```

Comment:
seems like we can at least `.drain()` here if not totally functionalize? not sure if flat_map+collect is smart enough to `Vec<Result<Vec>>` -> `Result<Vec>`

Reply:
i think it'd be something like...
```rust
 completed_ranges
    .drain()
    .flat_map(|(start_index, end_index)| {
        ...
        deshred()
            .map_err(|e| ...)
            .and_then(|dp| {
                deserialize(dp)
                    .map_err(|e| ...)
            })
    })
    .collect::<Result<Vec<_>, E>()
```

i know `collect()` can do shit like `Vec<Result<T, E>>` -> `Result<Vec<T>, E>`, hoping `flat_map()` can do the same type of transposition since it also operates on `IntoIterator`s. that is, goal is to go from `std::vec::Drain<<Result<std::vec::IntoIter<T>, E>>` -> Result<Vec<T>, E>` in one fell swoop

Reply:
Glad I asked, wasn't sure which item you were looking to drain.

In any case, don't think you can get it done with just `.flat_map()`; I found something in itertools tho:
https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.flatten_ok

Here's a playground if you want to tinker with it anymore:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=6b510b400d85a04f2a66054f662b4db4